pkg/csi_driver/multishare_controller.go

```diff
@@ -228,30 +216,121 @@ func (m *MultishareController) startAndWaitForInstanceDeleteOrShrink(ctx context
+	err = m.waitOnWorkflow(ctx, workflow)
 	if err != nil {
 		return status.Errorf(codes.Internal, "operation %q poll error: %v", workflow.opName, err)
 	}
 	return nil
 }
 
-func (m *MultishareController) ControllerExpandVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+func (m *MultishareController) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
 	// Handle higher level csi params validation, try locks
 	// Initiate share workflow by calling Multishare OpsManager functions
 	// Prepare and return csi response
-	// TODO
-	return nil, status.Error(codes.Unimplemented, "ControllerExpandVolume unsupported for multi share instances")
+
+	volumeId := req.GetVolumeId()
+	if len(volumeId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "ControllerExpandVolume volume ID must be provided")
+	}
+
+	reqBytes, err := getShareRequestCapacity(req.GetCapacityRange())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	_, project, location, instanceName, shareName, err := parseMultishareVolId(req.VolumeId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	klog.Infof("ControllerExpandVolume called for multishare with request %+v", req)
+	if acquired := m.volumeLocks.TryAcquire(volumeId); !acquired {
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volumeId)
+	}
+	defer m.volumeLocks.Release(volumeId)
+
+	share, err := m.cloud.File.GetShare(ctx, &file.Share{
+		Parent: &file.MultishareInstance{
+			Project:  project,
+			Location: location,
+			Name:     instanceName,
+		},
+		Name: shareName,
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+	if share == nil || file.IsNotFoundErr(err) {
+		return nil, status.Errorf(codes.Internal, "Couldn't find share with name %q", volumeId)
+	}
+
+	if share.CapacityBytes >= reqBytes {
+		klog.Infof("Controller expand volume succeeded for volume %v, existing size(bytes): %v", volumeId, share.CapacityBytes)
+		return &csi.ControllerExpandVolumeResponse{
+			CapacityBytes:         share.CapacityBytes,
+			NodeExpansionRequired: false,
+		}, nil
+	}
+
+	workflow, err := m.opsManager.expandInstanceAndStartWorkflow(ctx, share, reqBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.waitOnWorkflow(ctx, workflow)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "wait on operation %q failed with error: %v", workflow.opName, err)
+	}
+	klog.Infof("Wait for operation %s (type %s) completed", workflow.opName, workflow.opType.String())
+
+	switch workflow.opType {
+	case util.InstanceUpdate:
+		workflow, err = m.opsManager.startShareExpandWorkflowSafe(ctx, share, reqBytes)
+		if err != nil {
+			return nil, err
+		}
+	case util.ShareUpdate:
+		share, err = m.cloud.File.GetShare(ctx, share)
```

Comment:
we can make line 293 to 300 into a helper function getShareAndGenerateCSIControllerExpandVolumeResponse. reuse it in line 310-317